i3lock.c

```diff
@@ -645,6 +645,120 @@ void handle_screen_resize(void) {
     redraw_screen();
 }
 
+static int read_raw_image_native(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(&dest[y * stride], 1, width * 4, src);
+        count += n;
+        if (n < (size_t)(width * 4))
+            break;
+    }
+
+    return count;
+}
+
+static int read_raw_image_rgb(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    unsigned char *buf = malloc(width * 3);
+    if (buf == NULL)
+        return -1;
+
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(buf, 1, width * 3, src);
+        count += n;
+        if (n < (size_t)(width * 3))
+            break;
+
+        for (int x = 0; x < width; ++x) {
+            int idx = x * 3;
+            dest[y * stride + x] = 0 |
+                                   (buf[idx + 0] << 16) |
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
+        }
+    }
+
+    free(buf);
+    return count;
+}
+
+static cairo_surface_t *read_raw_image(const char *image_path, const char *image_raw_format) {
+    cairo_surface_t *img;
+
+    /* Parse format as <width>x<height>[:<format>] */
+    int w = -1, h = -1;
+    char format[] = "native";
+    char *fmtstring = NULL;
+    if (asprintf(&fmtstring, "%%ix%%i:%%%lds", sizeof(format) - 1) == -1) {
+        errx(EXIT_FAILURE, "asprintf() failed.");
+    }
+
+    if (sscanf(image_raw_format, fmtstring, &w, &h, format) == EOF || w == -1 || h == -1) {
+        fprintf(stderr, "Invalid image format \"%s\"\n", image_raw_format);
```

Comment:
Hm... I cannot add two separate comments to the same line? Anyway, the call to `sscanf` is still wrong:

```
RETURN VALUE
       On  success,  these functions return the number of input items successfully matched and assigned;
       this can be fewer than provided for, or even zero, in the event of an early matching failure.
```
Or is this why you check against `-1` instead of checking for `sscanf` to return 3?

Also... doesn't gcc warn against non-static format strings?

Reply:
We check w and h separately because those are the only required options, the format is optional (and defaults to native). It would be possible to check that sscanf returns at least 2, but I don't think it's a big deal. It would probably make sense though to check if w or h are less than 0, rather than checking if they're exactly -1, as you pointed out in the other comment.

Reply:
I think the >= 2 check actually sounds better than checking an individual error condition.